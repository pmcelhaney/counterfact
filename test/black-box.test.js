/* eslint-disable sonar/no-os-command-from-path */
/* eslint-disable sonar/os-command */
/* eslint-disable no-console */
/* eslint-disable n/no-process-env */
/* eslint-disable jest/no-restricted-matchers */
/* eslint-disable n/no-sync */
/* eslint-disable jest/no-hooks */
import { exec } from "node:child_process";
import fs from "node:fs";

import { afterAll, beforeAll, describe, expect, it } from "@jest/globals";
import fetch from "node-fetch";

const SERVER_START_WAIT_SECONDS = 10;

describe("black box test", () => {
  // eslint-disable-next-line init-declarations
  let counterfactProcess;

  beforeAll(async () => {
    counterfactProcess = exec(
      "node ./bin/counterfact.js ./openapi-example.yaml out",
      {
        env: { ...process.env, DEBUG: "counterfact:*" },
      },
    );

    counterfactProcess.stderr.pipe(process.stderr);
    counterfactProcess.stdout.pipe(process.stdout);

    // eslint-disable-next-line promise/avoid-new
    await new Promise((resolve) => {
      counterfactProcess.stdout.on("data", resolve);
    });

    // wait a few seconds to make sure the server is up
    // eslint-disable-next-line promise/avoid-new
    await new Promise((resolve) => {
      console.log(
        `waiting ${SERVER_START_WAIT_SECONDS} seconds for server to start`,
      );
      setTimeout(resolve, SERVER_START_WAIT_SECONDS * 1000);
    });
  }, 60_000);

  afterAll(() => {
    counterfactProcess.stdout.destroy();
    counterfactProcess.stderr.destroy();
    counterfactProcess.stdin.destroy();
    counterfactProcess.kill();
    counterfactProcess.unref();
  });

  it("launches the 'home page'", async () => {
    const response = await fetch("http://localhost:3100/counterfact/");

    expect(response.status).toBe(200);
  });

  it("creates a file for the /hello/kitty path", () => {
    expect(
      fs.readFileSync("./out/paths/hello/kitty.ts", "utf8"),
    ).toMatchSnapshot();
  });

  it("compiles kitty.ts", () => {
    expect(
      fs.readFileSync("./out/.cache/hello/kitty.js", "utf8"),
    ).toMatchSnapshot();
  });

  it("responds to a GET request", async () => {
    const response = await fetch("http://localhost:3100/hello/kitty");

    await expect(response.text()).resolves.toMatchSnapshot();
    expect(response.status).toBe(200);
  });
});
