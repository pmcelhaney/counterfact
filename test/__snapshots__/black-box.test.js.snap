// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`black box test compiles kitty.ts 1`] = `
""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GET = void 0;
const GET = ($) => {
    return $.response[200].random();
};
exports.GET = GET;
"
`;

exports[`black box test creates a file for the /hello/kitty path 1`] = `
"import type { HTTP_GET } from "../../types/paths/hello/kitty.types.js";

export const GET: HTTP_GET = ($) => {
  return $.response[200].random();
};
"
`;

exports[`black box test responds to a GET request 1`] = `"<img src="https://upload.wikimedia.org/wikipedia/en/0/05/Hello_kitty_character_portrait.png">"`;
